src/test/scala/com/databricks/labs/mosaic/test/MosaicSpatialQueryTest.scala

```diff
@@ -0,0 +1,136 @@
+    }
+
+    protected def testAllCodegen(testName: String, testTags: Tag*)(testFun: MosaicContext => Unit): Unit = {
+        for (geom <- geometryApis) {
+            for (is <- indexSystems) {
+                super.test(testName + s" (codegen) (${geom.name}, ${is.name})", testTags: _*)(
+                  withSQLConf(
+                    SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "false",
+                    SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
+                    SQLConf.CODEGEN_FACTORY_MODE.key -> CodegenObjectFactoryMode.CODEGEN_ONLY.toString
+                  ) {
+                      withMosaicConf(geom, is) {
+                          testFun
+                      }
+                  }
+                )
+            }
+        }
+    }
+
+    protected def testAllNoCodegen(testName: String, testTags: Tag*)(testFun: MosaicContext => Unit): Unit = {
+        for (geom <- geometryApis) {
+            for (is <- indexSystems) {
+                super.test(testName + s" (no codegen) (${geom.name}, ${is.name})", testTags: _*)(
+                  withSQLConf(
+                    SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "false",
+                    SQLConf.CODEGEN_FACTORY_MODE.key -> CodegenObjectFactoryMode.NO_CODEGEN.toString
+                  ) {
+                      withMosaicConf(geom, is) {
+                          testFun
+                      }
+                  }
+                )
+            }
+        }
+    }
+
+    def checkGeometryTopo(
+        mc: MosaicContext,
+        actualAnswer: DataFrame,
+        expectedAnswer: DataFrame,
+        geometryFieldName: String
+    ): Unit = {
+        MosaicSpatialQueryTest.checkGeometryTopo(mc, actualAnswer, expectedAnswer, geometryFieldName)
+    }
+
+}
+
+object MosaicSpatialQueryTest extends Assertions {
+    def checkGeometryTopo(
+        mc: MosaicContext,
+        actualAnswer: DataFrame,
+        expectedAnswer: DataFrame,
+        geometryFieldName: String
+    ): Unit = {
+        import mc.functions.st_aswkt
+
+        val actualGeoms = actualAnswer
+            .withColumn("answer_wkt", st_aswkt(col(geometryFieldName)))
+            .select(col("answer_wkt"))
+            .collect()
+            .map(_.getString(0))
+            .map(mc.getGeometryAPI.geometry(_, "WKT"))
+        val expectedGeoms = expectedAnswer
+            .withColumn("answer_wkt", st_aswkt(col(geometryFieldName)))
+            .select(col("answer_wkt"))
+            .collect()
+            .map(_.getString(0))
+            .map(mc.getGeometryAPI.geometry(_, "WKT"))
+
+        actualGeoms.zip(expectedGeoms).foreach { case (actualGeom, expectedGeom) =>
+            assert(actualGeom.equalsTopo(expectedGeom), s"$actualGeom did not topologically equal $expectedGeom")
+        }
+    }
+}
+
+trait MosaicHelper extends BeforeAndAfterEach { self: Suite =>
+    protected def withMosaicConf(geometry: GeometryAPI, indexSystem: IndexSystem)(f: MosaicContext => Unit): Unit = {
```

Comment:
technically this should say withMosaicContext since we are not really changing configuration.